src/SoapCore/MessageEncoder/SoapMessageEncoder.cs

```diff
@@ -146,35 +147,53 @@ public Task<Message> ReadMessageAsync(Stream stream, int maxSizeOfHeaders, strin
 			return Task.FromResult(message);
 		}
 
-		public virtual async Task WriteMessageAsync(Message message, PipeWriter pipeWriter)
+		public virtual async Task WriteMessageAsync(Message message, HttpContext httpContext)
 		{
 			if (message == null)
 			{
 				throw new ArgumentNullException(nameof(message));
 			}
 
+			if (httpContext == null)
+			{
+				throw new ArgumentNullException(nameof(httpContext));
+			}
+
+			var pipeWriter = httpContext.Response.BodyWriter;
 			if (pipeWriter == null)
 			{
 				throw new ArgumentNullException(nameof(pipeWriter));
 			}
 
 			ThrowIfMismatchedMessageVersion(message);
-
-			using var xmlTextWriter = XmlWriter.Create(pipeWriter.AsStream(true), new XmlWriterSettings
+	
+			//Custom string writer with custom encoding support
+			using (var stringWriter = new CustomStringWriter(_writeEncoding))
 			{
-				OmitXmlDeclaration = _optimizeWriteForUtf8 && _omitXmlDeclaration, //can only omit if utf-8
-				Indent = _indentXml,
-				Encoding = _writeEncoding,
-				CloseOutput = true,
-				CheckCharacters = _checkXmlCharacters
-			});
+				using (var xmlTextWriter = XmlWriter.Create(stringWriter, new XmlWriterSettings
+				{
+					OmitXmlDeclaration = _optimizeWriteForUtf8 && _omitXmlDeclaration, //can only omit if utf-8
+					Indent = _indentXml,
+					Encoding = _writeEncoding,
+					CloseOutput = true,
+					CheckCharacters = _checkXmlCharacters
+				}))
+				{
+					using var xmlWriter = XmlDictionaryWriter.CreateDictionaryWriter(xmlTextWriter);
+					message.WriteMessage(xmlWriter);
+					xmlWriter.WriteEndDocument();
+					xmlWriter.Flush();
+				}
 
-			using var xmlWriter = XmlDictionaryWriter.CreateDictionaryWriter(xmlTextWriter);
-			message.WriteMessage(xmlWriter);
-			xmlWriter.WriteEndDocument();
-			xmlWriter.Flush();
+				var data = stringWriter.ToString();
 
-			await pipeWriter.FlushAsync();
+				//Set Content-length in Response
+				httpContext.Response.ContentLength = data.Length;
```

Comment:
Do these legacy clients expect ContentLength to be the number of characters in the message, or the number of bytes?

Reply:
The legacy client worked with number of characters. Tested out with a unit test case and both the lengths were same. Leaving this change as is for now.